File: server/internal/store/sync_contract_vectors_test.go
package store_test

import (
	"encoding/json"
	"os"
	"slices"
	"sort"
	"testing"
	"time"

	"github.com/MohammadBnei/wird/server/internal/store"
	"github.com/MohammadBnei/wird/server/internal/testenv"
)

// contractPath is the second file both halves answer to, beside the set id
// vectors. The device suite reads these same rows from
// app/test/data/sync_contract_vectors_test.dart. Neither suite computes what
// it asserts, so the two cannot agree with themselves while disagreeing with
// each other — which is exactly how the set id shipped wrong.
const contractPath = "../../../docs/adr/0002-sync-contract-vectors.json"

type opBodyVector struct {
	Kind  string          `json:"kind"`
	Label string          `json:"label"`
	Body  json.RawMessage `json:"body"`
}

type opResultVector struct {
	Status    string `json:"status"`
	Landed    bool   `json:"landed"`
	Permanent bool   `json:"permanent"`
	Label     string `json:"label"`
}

type changeKindVector struct {
	Kind string                     `json:"kind"`
	Row  map[string]json.RawMessage `json:"row"`
}

type syncContract struct {
	OpBodies struct {
		Vectors []opBodyVector `json:"vectors"`
	} `json:"op_bodies"`
	OpResults struct {
		Vectors []opResultVector `json:"vectors"`
	} `json:"op_results"`
	ChangeKinds struct {
		Vectors []changeKindVector `json:"vectors"`
	} `json:"change_kinds"`
	ReadingOrders struct {
		Vectors []string `json:"vectors"`
		NotOne  string   `json:"not_a_reading_order"`
	} `json:"reading_orders"`
}

func contract(t *testing.T) syncContract {
	t.Helper()
	raw, err := os.ReadFile(contractPath)
	if err != nil {
		t.Fatalf("the shared sync contract is unreadable, so nothing checks the server against the device: %v", err)
	}
	var c syncContract
	if err := json.Unmarshal(raw, &c); err != nil {
		t.Fatalf("the shared sync contract is not readable as vectors: %v", err)
	}
	if len(c.OpBodies.Vectors) == 0 || len(c.OpResults.Vectors) == 0 ||
		len(c.ChangeKinds.Vectors) == 0 || len(c.ReadingOrders.Vectors) == 0 {
		t.Fatal("a section of the shared contract is empty, so this file passes without checking anything")
	}
	return c
}

// The failure: a field is renamed on one side, the decoder rejects the whole
// body because it disallows unknown fields, and the refusal is permanent — so
// the reader's prayer, note or understood aya is dead-lettered and never
// arrives. The bodies here are the device's own, checked in rather than built
// from these structs, so this reddens the moment the server stops answering
// to what the device actually sends.
func TestEveryOpBodyADeviceSendsIsAcceptedRatherThanRefusedForever(t *testing.T) {
	db, _ := testenv.Postgres(t)
	user := reader(t, db, "sub-op-bodies")

	for i, v := range contract(t).OpBodies.Vectors {
		op := store.Op{ClientOpID: opID(700 + i), Kind: v.Kind, Body: v.Body}
		results, err := db.Apply(t.Context(), user, []store.Op{op})
		if err != nil {
			t.Fatalf("%s (%s): %v", v.Kind, v.Label, err)
		}
		if results[0].Status != store.OpApplied {
			t.Errorf("%s (%s) came back %q: %s — the device spells this body one way and the server another, so the write is lost for good",
				v.Kind, v.Label, results[0].Status, results[0].Reason)
		}
	}
}

// The failure: a status word is renamed on the server, and the device reads
// the new word as transient — so it retries a write that can never land until
// the budget runs out, or parks one that could have landed. Both suites stay
// green, because each asserts against its own spelling.
func TestTheServerSpeaksTheOpOutcomeWordsTheDeviceParksAndRetriesOn(t *testing.T) {
	c := contract(t)
	byStatus := map[string]opResultVector{}
	for _, v := range c.OpResults.Vectors {
		byStatus[v.Status] = v
	}
	for word, constant := range map[string]string{
		"applied":   store.OpApplied,
		"duplicate": store.OpDuplicate,
		"refused":   store.OpRefused,
		"failed":    store.OpFailed,
	} {
		if _, ok := byStatus[constant]; !ok {
			t.Fatalf("the server answers %q where the shared contract knows only %v, and the device treats a word it has never heard of as worth retrying forever",
				constant, keys(byStatus))
		}
		if constant != word {
			t.Fatalf("the server answers %q where the contract names it %q", constant, word)
		}
	}

	// And the server really produces them. A word that is only a constant is
	// a word nothing checks.
	db, _ := testenv.Postgres(t)
	user := reader(t, db, "sub-outcomes")
	bodies := contract(t).OpBodies.Vectors
	first := store.Op{ClientOpID: opID(720), Kind: bodies[0].Kind, Body: bodies[0].Body}

	if got := status(t, db, user, first); got != store.OpApplied {
		t.Fatalf("a fresh write came back %q, not the contract's applied", got)
	}
	if got := status(t, db, user, first); got != store.OpDuplicate {
		t.Fatalf("a replayed write came back %q, not the contract's duplicate", got)
	}
	bad := store.Op{ClientOpID: opID(721), Kind: "ayah_understood", Body: json.RawMessage(`{"nope":1}`)}
	if got := status(t, db, user, bad); got != store.OpRefused {
		t.Fatalf("a body the decoder cannot read came back %q, not the contract's refused — the device will retry it until its budget runs out", got)
	}
}

// The failure: the server renames a kind, and the device's apply falls through
// to a default arm that returns zero. The device stops writing that table
// entirely and neither side raises anything. The kinds and their row keys are
// checked in, so a rename here has to be a rename there.
func TestTheChangeStreamEmitsOnlyTheKindsAndRowKeysTheDeviceKnowsHowToApply(t *testing.T) {
	c := contract(t)
	db, _ := testenv.Postgres(t)
	user := reader(t, db, "sub-kinds")

	// One row in every table the stream reads from, through the real ops.
	for i, v := range c.OpBodies.Vectors {
		if got := status(t, db, user, store.Op{ClientOpID: opID(740 + i), Kind: v.Kind, Body: v.Body}); got != store.OpApplied {
			t.Fatalf("seeding %s came back %q", v.Kind, got)
		}
	}

	page := pull(t, db, user, "")
	seen := map[string]json.RawMessage{}
	for _, change := range page.Changes {
		seen[change.Kind] = change.Row
	}
	want := map[string]bool{}
	for _, v := range c.ChangeKinds.Vectors {
		want[v.Kind] = true
		row, ok := seen[v.Kind]
		if !ok {
			t.Errorf("the stream carried no %q row, though the contract says the device applies one — either this kind is gone or nothing exercises it", v.Kind)
			continue
		}
		var got map[string]json.RawMessage
		if err := json.Unmarshal(row, &got); err != nil {
			t.Fatalf("%s row: %v", v.Kind, err)
		}
		if a, b := keys(got), keys(v.Row); !slices.Equal(a, b) {
			t.Errorf("the server builds a %s row out of %v where the device reads %v, so the device applies a row with fields missing", v.Kind, a, b)
		}
	}
	for kind := range seen {
		if !want[kind] {
			t.Errorf("the stream carries a %q the device has never heard of, and the device applies none of it", kind)
		}
	}
}

// The failure: a reading order the device cannot spell reaches the server,
// which stores it, derives a set id from it, and hands back a set no device
// recognises. The words are half of what names a set, so they are checked
// against the shared list rather than left to a column constraint.
func TestOnlyTheTwoReadingOrderWordsTheDeviceHasAreAccepted(t *testing.T) {
	c := contract(t)
	db, _ := testenv.Postgres(t)
	user := reader(t, db, "sub-orders")

	for i, order := range c.ReadingOrders.Vectors {
		op := prayedOp(opID(760+i), order, 1001, 1007, time.Now())
		if got := status(t, db, user, op); got != store.OpApplied {
			t.Errorf("the server refused %q, which is a word the device can and does send, so that reader's prayer is dead-lettered", order)
		}
	}

	stray := prayedOp(opID(780), c.ReadingOrders.NotOne, 1001, 1007, time.Now())
	if got := status(t, db, user, stray); got != store.OpRefused {
		t.Errorf("the server answered %q to reading order %q, so a set no device can name is now in the reader's history",
			got, c.ReadingOrders.NotOne)
	}
}

// status is land's cousin for the tests that are about the answer itself
// rather than about the write landing.
func status(t *testing.T, db *store.Store, user string, op store.Op) string {
	t.Helper()
	results, err := db.Apply(t.Context(), user, []store.Op{op})
	if err != nil {
		t.Fatalf("flush %s: %v", op.ClientOpID, err)
	}
	return results[0].Status
}

func prayedOp(clientOpID, readingOrder string, start, end int, at time.Time) store.Op {
	body, _ := json.Marshal(map[string]any{
		"set_id":        store.SetID(readingOrder, start, end),
		"start_ayah_id": start,
		"end_ayah_id":   end,
		"reading_order": readingOrder,
		"prayed_at":     at,
	})
	return store.Op{ClientOpID: clientOpID, Kind: "set_prayed", Body: body}
}

func keys[V any](m map[string]V) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: app/test/data/sync_contract_vectors_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:sqflite_common_ffi/sqflite_ffi.dart';
import 'package:wird/data/db.dart';
import 'package:wird/data/kept_repo.dart';
import 'package:wird/data/outbox.dart';
import 'package:wird/data/sets.dart';
import 'package:wird/data/sync.dart';
import 'package:wird/features/report/report.dart';

import '../corpus.dart';

/// The second file both halves answer to, beside the set id vectors. The
/// server suite reads these same rows from
/// server/internal/store/sync_contract_vectors_test.go. Neither suite computes
/// what it asserts, so the two cannot agree with themselves while disagreeing
/// with each other — which is exactly how the set id shipped wrong.
const _contractPath = '../docs/adr/0002-sync-contract-vectors.json';

Map<String, dynamic> _contract() {
  final file = File(_contractPath);
  expect(
    file.existsSync(),
    isTrue,
    reason: 'without the shared contract nothing checks the device against '
        'the server',
  );
  return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
}

List<Map<String, dynamic>> _section(String name) {
  final vectors =
      ((_contract()[name] as Map)['vectors'] as List).cast<Object>();
  expect(
    vectors,
    isNotEmpty,
    reason: 'an empty $name section would pass without checking anything',
  );
  return [for (final v in vectors) (v as Map).cast<String, dynamic>()];
}

/// One page of changes, served once. The push half is never reached: these
/// tests start with an empty queue.
class _Stream {
  _Stream._(this._server, this._changes) {
    unawaited(_serve());
  }

  static Future<_Stream> serving(List<Map<String, dynamic>> changes) async =>
      _Stream._(
        await HttpServer.bind(InternetAddress.loopbackIPv4, 0),
        changes,
      );

  final HttpServer _server;
  List<Map<String, dynamic>> _changes;

  Dio get dio => Dio(BaseOptions(baseUrl: 'http://127.0.0.1:${_server.port}'));

  Future<void> stop() => _server.close(force: true);

  Future<void> _serve() async {
    await for (final request in _server) {
      final page = {
        'changes': _changes,
        'cursor': 'seq:${_changes.length}',
        'more': false,
      };
      _changes = [];
      request.response
        ..headers.contentType = ContentType.json
        ..write(jsonEncode(page));
      unawaited(request.response.close());
    }
  }
}

Map<String, dynamic> _change(Map<String, dynamic> vector) {
  final row = (vector['row'] as Map).cast<String, dynamic>();
  return {
    'kind': vector['kind'],
    'id': '${row['id'] ?? 'prefs'}',
    'at': '2026-09-23T07:20:00Z',
    'row': row,
  };
}

void main() {
  late Database db;

  setUp(() async {
    db = await testCorpus();
    await ensureKeptTable(db);
    await db.delete('kept_items');
    await db.execute('DROP TABLE IF EXISTS sync_state');
  });

  // The failure: the device names a field one thing and the server's decoder
  // another. That decoder disallows unknown fields, so it refuses the whole
  // body — and a refusal is permanent, so the reader's note, prayer or
  // understood aya is parked and never arrives. The names here are checked in
  // rather than read off the device's own code, so a rename on either side of
  // the wire reddens this.
  test('every op the device builds carries the field names the server will '
      'accept, so no write is refused forever', () async {
    for (final vector in _section('op_bodies')) {
      if (vector['built_by_device'] != true) continue;
      final kind = vector['kind'] as String;
      await db.delete('outbox');

      switch (kind) {
        case 'ayah_understood':
          await markSetUnderstood(db, newOpId(), [1001, 1002]);
        case 'kept_upsert':
          await keep(db, kind: KeptKind.note, body: 'the mercy named twice');
        case 'kept_delete':
          final id = await keep(db, kind: KeptKind.note, body: 'to drop');
          await db.delete('outbox');
          await forget(db, id);
        case 'set_prayed':
          await recordSetPrayed(db, (await nextSet(db, ReadingOrder.nuzul))!);
        case 'prefs_set':
          await setReadingOrder(db, ReadingOrder.mushaf);
        case 'report_written':
          await sendReport(
            db,
            kind: ReportKind.bug,
            body: 'the audio stops at the end of the set',
            context: await reportContext(db, screen: 'prayer'),
          );
        default:
          fail('the shared contract carries an op kind "$kind" that no write '
              'path on this device builds, so nothing checks it');
      }

      final row = (await db.query('outbox')).single;
      expect(row['kind'], kind);
      final body = jsonDecode(row['body']! as String) as Map<String, dynamic>;
      expect(
        body.keys.toSet(),
        (vector['body'] as Map).keys.cast<String>().toSet(),
        reason: '${vector['label']}: the server refuses a body whose keys are '
            'not exactly these, and the refusal is permanent',
      );
    }
  });

  // The failure: the server renames one of its four answers. The device reads
  // an unknown word as transient, so it either retries a write that can never
  // land until the budget runs out and parks it anyway, or parks a write that
  // would have landed. Both suites stayed green through that until this file.
  test('the device parks and retries on the outcome words the server actually '
      'answers with', () {
    for (final vector in _section('op_results')) {
      final verdict = OpVerdict('op-1', vector['status'] as String);
      expect(
        verdict.landed,
        vector['landed'],
        reason: '${vector['label']}: the device disagrees about whether the '
            'server has this write',
      );
      expect(
        verdict.permanent,
        vector['permanent'],
        reason: '${vector['label']}: the device disagrees about whether this '
            'write can ever land',
      );
    }
  });

  // The failure: a second device prays a set and this one never counts it,
  // because the change arrives under a kind the apply switch has no arm for
  // and is dropped in silence. "The fourth prayer on this set" then reads
  // differently on the phone and on the tablet.
  test('every table the server streams changes for is written on this device, '
      'so a second device\'s prayers are counted here too', () async {
    final vectors = _section('change_kinds');
    final server = await _Stream.serving([for (final v in vectors) _change(v)]);
    addTearDown(server.stop);

    final report = await syncNow(db, SyncApi(server.dio));

    expect(report.unknownKinds, isEmpty);
    expect(
      report.applied,
      vectors.length,
      reason: 'a kind the server streams did not reach a local table',
    );
    expect(
      await prayersOnSet(db, '6cb8c394-ae9a-5092-a2ea-8f88749613c7'),
      1,
      reason: 'the prayer the other device made is not counted on this one',
    );
    expect(
      (await db.query('sets')).single['reading_order'],
      'mushaf',
      reason: 'the set the other device prayed did not reach this one',
    );
    expect((await db.query('ayah_understood')), hasLength(1));
    expect((await db.query('kept_items')), hasLength(1));
    expect((await db.query('user_prefs')).single['reading_order'], 'nuzul');
  });

  // The failure: the reader prays once, the server hands that prayer back on
  // the next pull, and the device counts it a second time — so the screen says
  // the fifth prayer on a set that has carried four. The ids are the device's
  // own, which is what makes the round trip idempotent, and the row shapes are
  // the ones the shared contract pins.
  test('a prayer the reader made comes back from the server and is counted '
      'twice', () async {
    final set = (await nextSet(db, ReadingOrder.nuzul))!;
    await recordSetPrayed(db, set);
    final body =
        jsonDecode((await db.query('outbox')).single['body']! as String)
            as Map<String, dynamic>;
    await db.delete('outbox');

    final server = await _Stream.serving([
      {
        'kind': 'sets',
        'id': body['set_id'],
        'at': body['prayed_at'],
        'row': {
          'id': body['set_id'],
          'ordinal': 1,
          'start_ayah_id': body['start_ayah_id'],
          'end_ayah_id': body['end_ayah_id'],
          'reading_order': body['reading_order'],
          'created_at': body['prayed_at'],
        },
      },
      {
        'kind': 'set_prayers',
        'id': body['id'],
        'at': body['prayed_at'],
        'row': {
          'id': body['id'],
          'set_id': body['set_id'],
          'prayer_name': '',
          'prayed_at': body['prayed_at'],
        },
      },
    ]);
    addTearDown(server.stop);

    await syncNow(db, SyncApi(server.dio));

    expect(await prayersOnSet(db, set.id), 1);
    expect((await db.query('sets')), hasLength(1));
  });

  // The failure: the server starts streaming a kind this build has never heard
  // of — a rename, or a table added after this app shipped — and the old
  // default arm returned zero, so the device stopped writing that table and
  // neither side said anything.
  test('a change kind this build cannot apply is raised rather than dropped in '
      'silence', () async {
    final server = await _Stream.serving([
      {
        'kind': 'root_known',
        'id': 'r1',
        'at': '2026-09-23T07:20:00Z',
        'row': {'root_letters': 'ح م د'},
      },
    ]);
    addTearDown(server.stop);

    await expectLater(
      syncNow(db, SyncApi(server.dio)),
      throwsA(isA<AssertionError>()),
    );
  });

  // The failure: the two halves spell a reading order differently. The word is
  // half of what a set id is derived from, so a third spelling is not a bad
  // value — it is a set no other device can name, and a prayer the server
  // refuses forever.
  test('the device can send only the reading-order words the server accepts',
      () {
    expect(
      {for (final order in ReadingOrder.values) order.name},
      ((_contract()['reading_orders'] as Map)['vectors'] as List)
          .cast<String>()
          .toSet(),
      reason: 'a word on one side of this that is not on the other is a set '
          'the other half cannot name',
    );
  });
}
